internal/component/discovery/target.go

```diff
@@ -0,0 +1,370 @@
+			default:
+				return fmt.Errorf("target::ConvertFrom: cannot convert value that can't be interfaced to (e.g. unexported struct field)")
+			}
+			labelSet[commonlabels.LabelName(k)] = commonlabels.LabelValue(strValue)
+		}
+		*t = NewTargetFromLabelSet(labelSet)
+		return nil
+	}
+	return fmt.Errorf("target: conversion from '%T' is not supported", src)
+}
+
+func (t Target) String() string {
+	s := make([]string, 0, t.Len())
+	t.ForEachLabel(func(key string, value string) bool {
+		s = append(s, fmt.Sprintf("%q=%q", key, value))
+		return true
+	})
+	slices.Sort(s)
+	return fmt.Sprintf("{%s}", strings.Join(s, ", "))
+}
+
+// Equals implements equality.CustomEquality. Works only with pointers.
+func (t Target) Equals(other any) bool {
+	if ot, ok := other.(*Target); ok {
+		return t.EqualsTarget(ot)
+	}
+	return false
+}
+
+func (t Target) EqualsTarget(other *Target) bool {
+	if t.Len() != other.Len() {
+		return false
+	}
+	finished := t.ForEachLabel(func(key string, value string) bool {
+		otherValue, ok := other.Get(key)
+		if !ok || otherValue != value {
+			return false
+		}
+		return true
+	})
+	return finished
+}
+
+func (t Target) NonMetaLabelsHash() uint64 {
+	return t.HashLabelsWithPredicate(func(key string) bool {
+		return !strings.HasPrefix(key, commonlabels.MetaLabelPrefix)
+	})
+}
+
+func (t Target) SpecificLabelsHash(labelNames []string) uint64 {
+	return t.HashLabelsWithPredicate(func(key string) bool {
+		return slices.Contains(labelNames, key)
+	})
+}
+
+func (t Target) HashLabelsWithPredicate(pred func(key string) bool) uint64 {
+	// For hash to be deterministic, we need labels order to be deterministic too. Figure this out first.
+	labelsInOrder := stringSlicesPool.Get().([]string)
+	defer stringSlicesPool.Put(labelsInOrder[:])
+	t.ForEachLabel(func(key string, value string) bool {
+		if pred(key) {
+			labelsInOrder = append(labelsInOrder, key)
+		}
+		return true
+	})
+	slices.Sort(labelsInOrder)
+	return t.hashLabelsInOrder(labelsInOrder)
+}
+
+func (t Target) groupLabelsHash() uint64 {
+	// For hash to be deterministic, we need labels order to be deterministic too. Figure this out first.
+	labelsInOrder := stringSlicesPool.Get().([]string)
+	defer stringSlicesPool.Put(labelsInOrder[:])
+
+	for name := range t.group {
+		labelsInOrder = append(labelsInOrder, string(name))
+	}
+	slices.Sort(labelsInOrder)
+	return t.hashLabelsInOrder(labelsInOrder)
+}
+
+func (t Target) hashLabelsInOrder(order []string) uint64 {
+	// This optimisation is adapted from prometheus/model/labels.
+	// Use xxhash.Sum64(b) for fast path as it's faster.
+	b := make([]byte, 0, 1024)
+	mustGet := func(label string) string {
+		val, ok := t.Get(label)
+		if !ok { // should never happen as Target is immutable
+			panic("label concurrently modified - this is a bug - please report an issue")
+		}
+		return val
+	}
+
+	for i, key := range order {
+		value := mustGet(key)
+		if len(b)+len(key)+len(value)+2 >= cap(b) {
```

Comment:
I assume the +2 is from seps?

Reply:
Yep. All of this code comes from Prometheus: https://github.com/grafana/prometheus/blob/staleness_disabling_v0.54.1/model/labels/labels.go#L75

And we must match it because we don't want the rollout to rehash all targets differently. There are tests to guarantee this too.

Reply:
The the funcs/files that come from Prometheus can we document that and that it cant be changed for backwards compat?